src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```diff
@@ -742,35 +743,163 @@ public override ClaimsPrincipal ValidateToken(string token, TokenValidationParam
-
+            ClaimsPrincipal claimsPrincipal;
+            SecurityToken signatureValidatedToken;
             if (tokenParts.Length == JwtConstants.JweSegmentCount)
             {
                 var jwtToken = ReadJwtToken(token);
                 var decryptedJwt = DecryptToken(jwtToken, validationParameters);
-                var innerToken = ValidateSignature(decryptedJwt, validationParameters);
-                jwtToken.InnerToken = innerToken;
-                signatureValidatedToken = jwtToken;
-                claimsPrincipal = ValidateTokenPayload(innerToken, validationParameters);
+                claimsPrincipal = ValidateToken(decryptedJwt, jwtToken, validationParameters, out signatureValidatedToken);
             }
             else
             {
-                signatureValidatedToken = ValidateSignature(token, validationParameters);
-                claimsPrincipal = ValidateTokenPayload(signatureValidatedToken as JwtSecurityToken, validationParameters);
+                claimsPrincipal = ValidateToken(token, null, validationParameters, out signatureValidatedToken);
             }
 
             validatedToken = signatureValidatedToken;
             return claimsPrincipal;
         }
 
+        /// <summary>
+        ///  Private method for token validation, responsible for:
+        ///  (1) Obtaining a configuration from the <see cref="TokenValidationParameters.ConfigurationManager"/>.
+        ///  (2) Revalidating using the Last Known Good Configuration (if present), and obtaining a refreshed configuration (if necessary) and revalidating using it.
+        /// </summary>
+        /// <param name="token">The JWS string, or the decrypted token if the token is a JWE.</param>
+        /// <param name="outerToken">If the token being validated is a JWE, this is the <see cref="JwtSecurityToken"/> that represents the outer token.
+        ///  If the token is a JWS, the value of this parameter is null.
+        /// </param>
+        /// <param name="validationParameters">The <see cref="TokenValidationParameters"/> to be used for validation.</param>
+        /// <param name="signatureValidatedToken">The <see cref="JwtSecurityToken"/> that was validated.</param>
+        /// <returns> A <see cref="ClaimsPrincipal"/> from the JWT. Does not include claims found in the JWT header.</returns>
+        private ClaimsPrincipal ValidateToken(string token, JwtSecurityToken outerToken, TokenValidationParameters validationParameters, out SecurityToken signatureValidatedToken)
+        {
+            BaseConfiguration currentConfiguration = null;
+            if (validationParameters.ConfigurationManager != null)
+            {
+                try
+                {
+                    currentConfiguration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
```

Comment:
>Exception ex [](http://example.com/codeflow?start=23&length=12)

Is there a set of library specific exceptoins we can catch/swallow here? The comment makes sense for a known exception type but this will catch everything. Logging all exceptions makes sense to me but swallowing all exceptions seems dangerous.

Reply:
@TimHannMSFT any exceptions that are about to bring down the appdomain will fail again shortly, no?
What would you suggest? At one point in time there was an effort to fail fast on certain exceptions, are you suggesting we have such a list?
This model is to ensure for any reason if Configuration was not able to be obtained, we can move forward as the artifacts needed for validation may be in TVP.

Reply:
It depends and fatal system exceptions can have goofy things occur if you keep carrying on as if nothing happened. I don't think anyone can confidently say how the system will behave if it's out of memory...perhpas garbage collection occurs at just the right time and now there's more but the state is corrupted, I don't think the code can or should try to reason about this, fail fast.

My suggestion is catch what you know and throw what you don't. Absolutely log everything but don't try and pretend your code is smart enough to handle all contigencies.

To my mind there are three cases:
1. System exceptions - the system is now in a state you can't really control or predict, you should just bubble this up after logging.
2. Some underlying library has thrown an exception you didn't account for in your code and don't know how to handle - your assumptions are wrong and it's no longer safe to continue processing because you'd definitionally in  state you don't understand.
3. An exceptional case you know about - here you should catch the excepiton by the known type and handle it in the way you know how.

There are exceptions to the rule of course but I think we should at least have a long comment explaining why we'd make an exception in a library. (If you're running a service that "can't" fail at some critical point maybe it makes sense to try and do everything you can to hang on but IMO we don't hae the context to make those sorts of calls at our layer...fail fast)

Reply:
Also meant to acknowledge that at least some of the sytem exceptions aren't actually caught by this pattern for just this reason (IIRC this was added later in .NET) So for some of (1) we're covered but I'm not sure about all cases and (2) still is a concern.

Reply:
@TimHannMSFT even logging can be a problem. Last i checked with the .net team is that most exceptions that would cause major issues to ignore we won't see, but we should investigate how the runtime handles system exceptions.